rfcs/2018-11-23-annotations.md

```diff
@@ -0,0 +1,406 @@
+# RFC: Annotations
+
+_In which we propose the concept of and a draft API for ‘Annotations’,
+a new way for core and plugins to represent additional information about
+regions of a document._
+
+## Summary
+
+Add ‘annotations’, a new concept and API for the representation of
+transient document state. This provides a flexible mechanism for
+representing a wide variety of core and plugin information to the
+frontend, without using the text styling API.
+
+This API will be the basis for a wide variety of future features,
+including things like:
+
+- compiler warnings / lints / diagnostic messages
+- displaying other user's cursors during multi-user editing
+- clickable external links
+- version control status
+- breakpoints
+
+Additionally, it will be used to represent the user's _selections_ and
+_find highlights_.
+
+The text styling API will not be materially changed as a result of this
+proposal, although it will be slightly simplified.
+
+## Overview
+
+_Annotations_ are a very simple abstraction. The basic idea is to
+generalize the common pattern of having some _type_ of data associated
+with some set of document _regions_. Annotations themselves have types,
+and these types determine the expected structure of a given annotation's
+data, or payload. How (and whether) to handle a given type of annotation
+is left up to the frontend, although suggestions or guidance may be
+provided.
+
+We will add an `annotations` field to the update protocol. This field
+will be a list of sets of annotations, which will represent _transient
+state_ scoped to the current view. Each annotations set will include
+a type identifier and two arrays, one of byte ranges and one of
+corresponding payloads. In sort-of JSON, this might look like:
+
+```javascript
+// in the update object:
+{
+    "ops": [Op]?,
+    "annotations": [AnnotationSet]?,
+    // other fields
+}
+
+// An AnnotationSet:
+{
+    "type": String,
+    "ranges": [[Number, Number]],
```

Comment:
My understanding is that we don't use byte ranges in the update protocol. I believe when we discussed this, we were considering making the endpoints of a range be (visual) line number and offset-within-line, where the latter could reasonably be utf-8 offset, utf-16, or possibly both.